File: python/387.first-unique-character-in-a-string.py
#
# @lc app=leetcode.cn id=387 lang=python3
#
# [387] 字符串中的第一个唯一字符
#
# https://leetcode-cn.com/problems/first-unique-character-in-a-string/description/
#
# algorithms
# Easy (51.19%)
# Total Accepted:    173.1K
# Total Submissions: 334.5K
# Testcase Example:  '"leetcode"'
#
# 给定一个字符串，找到它的第一个不重复的字符，并返回它的索引。如果不存在，则返回 -1。
# 
# 
# 
# 示例：
# 
# s = "leetcode"
# 返回 0
# 
# s = "loveleetcode"
# 返回 2
# 
# 
# 
# 
# 提示：你可以假定该字符串只包含小写字母。
# 
#
class Solution:
    def firstUniqChar(self, s: str) -> int:
        d = {}
        for l in s:
            if l in d:
                d[l] += 1
            else:
                d[l] = 1
        for i, l in enumerate(s):
            if d[l] == 1:
                return i
        return -1

File: cpp/387.first-unique-character-in-a-string.cpp
/*
 * @lc app=leetcode.cn id=387 lang=cpp
 *
 * [387] 字符串中的第一个唯一字符
 *
 * https://leetcode-cn.com/problems/first-unique-character-in-a-string/description/
 *
 * algorithms
 * Easy (53.53%)
 * Total Accepted:    277.4K
 * Total Submissions: 509.4K
 * Testcase Example:  '"leetcode"'
 *
 * 给定一个字符串 s ，找到 它的第一个不重复的字符，并返回它的索引 。如果不存在，则返回 -1 。
 * 
 * 
 * 
 * 示例 1：
 * 
 * 
 * 输入: s = "leetcode"
 * 输出: 0
 * 
 * 
 * 示例 2:
 * 
 * 
 * 输入: s = "loveleetcode"
 * 输出: 2
 * 
 * 
 * 示例 3:
 * 
 * 
 * 输入: s = "aabb"
 * 输出: -1
 * 
 * 
 * 
 * 
 * 提示:
 * 
 * 
 * 1 <= s.length <= 10^5
 * s 只包含小写字母
 * 
 * 
 */
class Solution {
public:
    int firstUniqChar(string s) {
        unordered_map<int, int> hash;
        for (char ch: s){
            ++hash[ch];
        }

        for (int i=0; i<s.size(); i++){
            if (hash[s[i]] == 1){
                return i;
            }
        }

        return -1;

    }
};
